manticore/core/workspace.py

```diff
@@ -0,0 +1,458 @@
+        path = os.path.join(self.uri, key)
+        os.remove(path)
+
+    def ls(self, glob_str):
+        '''
+        Return just the filenames that match `glob_str` inside the store directory.
+
+        :param glob_str:
+        :return:
+        '''
+        path = os.path.join(self.uri, glob_str)
+        return map(lambda s: os.path.split(s)[1], glob.glob(path))
+
+
+class RedisStore(Store):
+    '''
+    A redis-backed Manticore workspace
+    '''
+    def __init__(self, uri=None):
+        '''
+        :param uri: A url for redis
+        '''
+
+        # Local import to not create an explicit dependency
+        import redis
+
+        hostname, port = uri.split(':')
+        self._client = redis.StrictRedis(host=hostname, port=int(port), db=0)
+
+        super(RedisStore, self).__init__(uri)
+
+    def save_value(self, key, value):
+        '''
+        Save an arbitrary, serializable `value` under `key`.
+
+        :param str key: A string identifier under which to store the value.
+        :param value: A serializable value
+        :return:
+        '''
+        return self._client.set(key, value)
+
+    def load_value(self, key):
+        '''
+        Load an arbitrary value identified by `key`.
+
+        :param str key: The key that identifies the value
+        :return: The loaded value
+        '''
+        return self._client.get(key)
+
+    def rm(self, key):
+        self._client.delete(key)
+
+    def ls(self, glob_str):
+        return self._client.keys(glob_str)
+
+def _create_store(desc):
+    '''
+    Create a :class:`~manticore.core.workspace.Store` instance depending on the descriptor.
+
+    Valid descriptors:
+      fs:<path>
+      redis:<hostname>:<port>
+
+    :param str desc: Store descriptor
+    :return: Store instance
+    '''
+    type, uri = ('fs', None) if desc is None else desc.split(':', 1)
+
+    if type == 'fs':
+        return FilesystemStore(uri)
+    elif type == 'redis':
+        return RedisStore(uri)
+    else:
+        raise NotImplementedError("Storage type '%s' not supported.", type)
+
+class Workspace(object):
+    '''
+    A workspace maintains a list of states to run and assigns them IDs.
+    '''
+
+    def __init__(self, lock, desc=None):
+        self._store = _create_store(desc)
+        self._serializer = PickleSerializer()
+        self._last_id = 0
+        self._lock = lock
+        self._prefix = 'state_'
+        self._suffix = '.pkl'
+
+    def try_loading_workspace(self):
+        state_names = self._store.ls('{}*'.format(self._prefix))
+        def get_state_id(name):
+            return int(name[len(self._prefix):-len(self._suffix)], 16)
+        state_ids = map(get_state_id, state_names)
+
+        if not state_ids:
+            return []
+
+        self._last_id = max(state_ids) + 1
+
+        return state_ids
+
+    def _get_id(self):
+        '''
+        Get a unique state id.
+
+        :rtype: int
+        '''
+        try:
+            self._lock.acquire()
```

Comment:
I wonder if we need to use the  global executor lock for this or if there is a locking mechanism more close to the workspace implementation..